test/framework/exec/docker.go

```diff
@@ -0,0 +1,360 @@
+		AttachStderr: true,
+	}
+	hostConfig := container.HostConfig{
+		SecurityOpt:  []string{"seccomp=unconfined"}, // ignore seccomp
+		Binds:        bindings,
+		NetworkMode:  container.NetworkMode(net),
+		PortBindings: nat.PortMap{},
+	}
+	networkConfig := network.NetworkingConfig{}
+
+	envVars := []string{}
+	for key, val := range env {
+		envVars = append(envVars, fmt.Sprintf("%s=%s", key, val))
+	}
+
+	// pass proxy environment variables to be used by node's docker daemon
+	proxyDetails := getProxyDetails()
+	for key, val := range proxyDetails.Envs {
+		envVars = append(envVars, fmt.Sprintf("%s=%s", key, val))
+	}
+	containerConfig.Env = envVars
+
+	if usernsRemap(ctx) {
+		// We need this argument in order to make this command work
+		// in systems that have userns-remap enabled on the docker daemon
+		hostConfig.UsernsMode = "host"
+	}
+
+	// Make sure we have the image
+	err = PullDockerImage(ctx, image)
+	if err != nil {
+		return err
+	}
+
+	// Create the container using our settings
+	resp, err := docker.ContainerCreate(
+		ctx,
+		&containerConfig,
+		&hostConfig,
+		&networkConfig,
+		nil,
+		"",
+	)
+	if err != nil {
+		return err
+	}
+
+	// Read out any output from the container
+	attachOpts := types.ContainerAttachOptions{
+		Stream: true,
+		Stdin:  false,
+		Stdout: true,
+		Stderr: true,
+	}
+
+	// Attach to the container so we can capture the output
+	containerOutput, err := docker.ContainerAttach(ctx, resp.ID, attachOpts)
+	if err != nil {
+		return err
+	}
+
+	// Actually start the container
+	if err := docker.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); err != nil {
+		return err
+	}
+
+	outputErrors := make(chan error)
+	go func() {
+		// Send the output to the host file
+		_, err = io.Copy(output, containerOutput.Reader)
+		outputErrors <- err
+	}()
+	defer containerOutput.Close()
+
+	select {
+	case err := <-outputErrors:
+		if err != nil {
+			return err
+		}
+	case <-ctx.Done():
+		return err
+	}
+
+	// Wait for the run to complete
+	statusCh, errCh := docker.ContainerWait(ctx, resp.ID, container.WaitConditionNotRunning)
+	select {
+	case err := <-errCh:
+		if err != nil {
+			return err
+		}
+	case <-statusCh:
+	case <-ctx.Done():
+		return err
+	}
+
+	return nil
+}
+
+// proxyDetails contains proxy settings discovered on the host.
+type proxyDetails struct {
+	Envs map[string]string
+}
+
+// getProxyDetails returns a struct with the host environment proxy settings
+// that should be passed to the nodes.
+func getProxyDetails() *proxyDetails {
+	var val string
+	details := proxyDetails{Envs: make(map[string]string)}
+	proxyEnvs := []string{httpProxy, httpsProxy, noProxy}
+
+	for _, name := range proxyEnvs {
+		val = os.Getenv(name)
+		if val != "" {
+			details.Envs[name] = val
+			details.Envs[strings.ToLower(name)] = val
+		} else {
+			val = os.Getenv(strings.ToLower(name))
+			if val != "" {
+				details.Envs[name] = val
+				details.Envs[strings.ToLower(name)] = val
+			}
+		}
+	}
```

Comment:
```suggestion
	for _, name := range proxyEnvs {
		val = os.Getenv(name)
		if val == "" {
			val = os.Getenv(strings.ToLower(name))
		}
		if val == "" {
			continue
		}

		details.Envs[name] = val
		details.Envs[strings.ToLower(name)] = val
	}
```
Just an idea

Reply:
And a good idea. Gets rid of duplication and I think makes it easier to read. Updated.